wp/plugin/lib/equality.ml

```diff
@@ -0,0 +1,387 @@
+    sub_exp map exp1 exp2 >>=
+    fun (map, exp1) ->
+    if Exp.equal exp1 exp2 then
+      Some map
+    else None
+  | _, _ -> None
+
+(* Compares two things:
+   - jmp1 and jmp2 match in structure
+   - jmp1 and jmp2 match in all Exp.ts contained within *)
+let compare_jmp (jmp1 : jmp term) (jmp2 : jmp term) (map : varToVarMap) :
+  varToVarMap option =
+  match Jmp.kind jmp1, Jmp.kind jmp2 with
+  | Goto label1, Goto label2 -> compare_lbl label1 label2 map
+  | Call call1, Call call2 ->
+    begin
+      (* TODO assert equality *)
+      let target1, target2 = Call.target call1, Call.target call2 in
+      compare_lbl target1 target2 map >>=
+      fun map ->
+      match Call.return call1, Call.return call2 with
+      | Some label1, Some label2 -> compare_lbl label1 label2 map
+      | None, None -> Some map
+      | _, _ -> None
+    end
+  | Ret label1, Ret label2 -> compare_lbl label1 label2 map
+  | Int (code1, _tid1), Int (code2, _tid2) -> if code1 = code2 then Some map else None
+  | _, _ -> None
+
+(* There should be no phi terms. Fail if there are. *)
+let compare_phis (phis1 : (phi term) list) (phis2 : (phi term) list)
+    (map : varToVarMap) : varToVarMap option =
+  if List.length phis1 > 0 || List.length phis2 > 0 then None
+  else Some map
+
+(* Check that all defs in two lists match in expression and structure but not TID. *)
+let compare_defs (defs1 : (def term) list)
+    (defs2 : (def term) list) (map : varToVarMap) : varToVarMap option =
+  match List.zip defs1 defs2 with
+  | Core_kernel.List.Or_unequal_lengths.Unequal_lengths -> None
+  | Core_kernel.List.Or_unequal_lengths.Ok z ->
+    List.fold z ~init:(Some map)
+      ~f:(fun (map_wrap) (d1, d2) ->
+          match map_wrap with
+          | None -> None
+          | Some map -> compare_def d1 d2 map)
+
+(* Check that all jmps in a list match in expression and structure but not TID. *)
+let compare_jmps (jmps1 : (jmp term) list) (jmps2 : (jmp term) list)
+    (map : varToVarMap) : varToVarMap option =
+  match List.zip jmps1 jmps2 with
+  | Core_kernel.List.Or_unequal_lengths.Unequal_lengths -> None
+  | Core_kernel.List.Or_unequal_lengths.Ok z ->
+    List.fold z ~init:(Some map)
+      ~f:(fun (map_wrap) (j1, j2) ->
+          match map_wrap with
+          | Some map -> compare_jmp j1 j2 map
+          | None -> None)
+
+(* Compares everything about a block EXCEPT tids, which will be done later. *)
+let compare_block (blk1 : blk term) (blk2 : blk term) map =
+  compare_defs (Term.enum def_t blk1 |> Sequence.to_list)
+    (Term.enum def_t blk2 |> Sequence.to_list) map >>=
+  fun map -> compare_jmps (Term.enum jmp_t blk1 |> Sequence.to_list)
+    (Term.enum jmp_t blk2 |> Sequence.to_list) map >>=
+  fun map -> compare_phis (Term.enum phi_t blk1 |> Sequence.to_list)
+    (Term.enum phi_t blk2 |> Sequence.to_list) map
+
+(* Iterate through all reachable nodes in each cfg
+ *  and generate a map from tid to blk.*)
+let get_node_maps
```

Comment:
Since the two cfgs and maps are built entirely independently, couldn't one make a `get_node_map : Ir.t -> blk TidMap.t` function that you call twice instead?